csi/src/dev.rs

```diff
@@ -0,0 +1,175 @@
+//! Definition of traits required for attaching and detaching devices.
+//! Note that (unfortunately) the attach and detach operations are not
+//! quite symmetric. This is because it is not possible to discover all
+//! the information contained in a device URI from the entries in udev.
+//! Hence there are separate traits for attach and detach that each
+//! device type must implement.
+//!
+//! Attaching a device is performed as follows:
+//! ```ignore
+//!     let uri = "iscsi://192.168.0.20:3260/iqn.2019-05.io.openebs:nexus-11111111-0000-0000-0000-000000000000/0";
+//!     let device = Device::parse(uri)?;
+//!     if let Some(path) = device.find().await? {
+//!         // device already attached
+//!     } else {
+//!         // attach the device
+//!         device.attach().await?;
+//!         // wait for it to show up in udev and obtain the path
+//!         let path = Device::wait_for_device(device, 10).await?;
+//!     }
+//! ```
+//!
+//! Detaching a device is performed via:
+//! ```ignore
+//!     let uuid = Uuid::parse_str(&volume_id)?;
+//!     if let Some(device) = Device::lookup(&uuid).await? {
+//!         device.detach().await?;
+//!     }
+//! ```
+
+use std::{convert::TryFrom, time::Duration};
+
+use tokio::time::delay_for;
+use udev::Enumerator;
+use url::Url;
+use uuid::Uuid;
+
+mod iscsi;
+mod nbd;
+mod nvmf;
+mod util;
+
+const NVME_NQN_PREFIX: &str = "nqn.2019-05.io.openebs";
+
+pub use crate::error::DeviceError;
+pub type DeviceName = String;
+
+#[tonic::async_trait]
+pub trait Attach: Sync + Send {
+    async fn attach(&self) -> Result<(), DeviceError>;
+    async fn find(&self) -> Result<Option<DeviceName>, DeviceError>;
+}
+
+#[tonic::async_trait]
+pub trait Detach: Sync + Send {
+    async fn detach(&self) -> Result<(), DeviceError>;
+    fn devname(&self) -> DeviceName;
+}
+
+pub struct Device;
+
+impl Device {
+    /// Main dispatch function for parsing URIs in order
+    /// to obtain a device implementing the Attach trait.
+    pub fn parse(uri: &str) -> Result<Box<dyn Attach>, DeviceError> {
+        let url = Url::parse(uri).map_err(|error| error.to_string())?;
+        match url.scheme() {
+            "file" => Ok(Box::new(nbd::Nbd::try_from(&url)?)),
+            "iscsi" => Ok(Box::new(iscsi::IscsiAttach::try_from(&url)?)),
+            "nvmf" => Ok(Box::new(nvmf::NvmfAttach::try_from(&url)?)),
+            "nbd" => Ok(Box::new(nbd::Nbd::try_from(&url)?)),
+            scheme => Err(DeviceError::from(format!(
+                "unsupported device scheme: {}",
+                scheme
+            ))),
+        }
+    }
+
+    /// Lookup an existing device in udev matching the given UUID
+    /// to obtain a device implementing the Detach trait.
+    pub async fn lookup(
```

Comment:
I have difficultly understanding the difference between lookup and the find method defined with the trait, the both amount to the same thing. Why cant I attach and then do a lookup instead of doing an attach and then find?

Reply:
They do serve different purposes. find() is part of the trait while lookup() is not. lookup() is needed when you don't have an object or a URI with which to create one - you only have a UUID. This is inteded to be used when detaching.  find() is intended to be called once you have an object (obtained from URI) in order to attach. It's job is to tell you where to find the device referred to in the URI. There are also different implementations of find(), which need not use udev at all (such as  nbd - where find() doesn't involve a udev lookup  and does not implement detach at all).